---
title:  "Open source alternatives"
date:   2023-05-24 09:20:11 -0400
last_modified_at: 2023-05-26 16:20:02 -0400
categories: BioNumerics genomic-epidemiology open-source

---

BioNumerics is [sunsetting in December 2024](https://www.applied-maths.com/news/bionumerics-phasing-out).
This is an issue especially in the PulseNet community where the network relies on this software to do many things including
genome assembly,
molecular typing,
and MLST.

Although these aren't recommendations from CDC, I wanted to list some viable alternatives.

## General

| software or site | Workflow language | Containerization | Execution backend | notes |
|------------------|-------------------|-----------------|---------------|-------|
| [Terra.bio](https://terra.bio/)        | WDL | Docker | Google | [SARS-CoV-2 protocols](https://www.protocols.io/workspaces/toastpublic/publications), [terra.bio documentation](https://terra.bio/resources/getting-started/) |
| [Bactopia](https://bactopia.github.io/) | Nextflow | Conda, Docker, or Singularity | local, HPC, Azure, AWS, GCP, [more](https://www.nextflow.io/docs/latest/executor.html)| [documentation](https://bactopia.github.io/) |
| [StaPH-B Toolkit](https://staphb.org/staphb_toolkit/) | Nextflow | Docker or Singularity | local, HPC, Azure, AWS, GCP, [more](https://www.nextflow.io/docs/latest/executor.html) | [documentation](https://staphb.org/staphb_toolkit/) |
| [GalaxyTrakr](https://galaxytrakr.org/) | Galaxy | Galaxy environments | FDA cloud | [Documentation](https://galaxyproject.org/support/) |
| [IRIDA](https://sfu.irida.ca/) | Galaxy | Galaxy environments | PHAC cloud | [Documentation](http://www.irida.ca/getting-started/) |
| [EDGE](https://edgebioinformatics.org/) | Unsure | EDGE modules | local or on their website | [documentation](https://edgebioinformatics.org/) |
| [Easy Genomics](https://www.easygenomics.org/) |

## Specific

Usually, a person just wants to replace one analysis such as genome assembly and so here are some options.

| analysis | BioNumerics software | alternative |
|----------|----------------------|-------------|
| Genome assembly (Illumina) | SPAdes        | [Shovill with SPAdes](https://github.com/tseemann/shovill) |
| Genome assembly (ONT) | unsure | [Dragonflye](https://github.com/rpetit3/dragonflye) |
| Serotyping (Salmonella) | SeqSero2 | [SeqSero2](http://www.denglab.info/SeqSero), [SISTR](https://lfz.corefacility.ca/sistr-app/) |
| Genotyping | bespoke | [ReMatCh](http://github.com/B-UMMI/ReMatCh), [Abricate](https://github.com/tseemann/abricate/issues), [SRST2](http://github.com/katholt/srst2) |
| 7-gene MLST | bespoke | Usually [_mlst_](https://github.com/tseemann/mlst) but this is a more comprehensive [comparison manuscript](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5610716/) |
| cg/wgMLST | bespoke | [ChewBBACA](https://chewbbaca.online/), [ColorID](https://github.com/hcdenbakker/colorid), [EToKi](https://github.com/zheminzhou/EToKi). See also my [detailed usage post]({% post_url 2023-04-09-wgMLST %}). |
| SNP       | bespoke | [Lyve-SET](https://github.com/lskatz/lyve-SET), [CFSAN SNP-Pipeline](https://snp-pipeline.readthedocs.io/en/latest/), [SNVPhyl](https://github.com/phac-nml/irida/tree/master). See also our [comparison between SNP pipelines](https://www.frontiersin.org/articles/10.3389/fmicb.2017.00375/full). |
| Quality Assurance/Quality Checks (QA/QC) | Bespoke | [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) |
| Contamination detection | Bespoke | [Kraken](https://ccb.jhu.edu/software/kraken/), [MIDAS](https://github.com/snayfach/MIDAS) |
| Virulence finder | Bespoke | [ResFinder](https://cge.cbs.dtu.dk/services/ResFinder/), [AMRFinderPlus](https://github.com/ncbi/amr) |
| Gene prediction and annotation | Bespoke | [Prokka](http://github.com/tseemann/prokka), [Bakta](https://github.com/oschwengers/bakta#database) |
| Genome alignment | Bespoke | [Harvest](http://harvest.readthedocs.io/en/latest/) |
| Gene alignment | Bespoke | [MEGA](https://www.megasoftware.net/) |
| Phylogeny | Bespoke | [MEGA (for small alignments)](https://www.megasoftware.net/), [Harvest (for whole genomes)](http://harvest.readthedocs.io/en/latest) |

Many, many more tools can be found in this [2018 paper](https://www.sciencedirect.com/science/article/pii/S1198743X17307097?via%3Dihub).

## Acknowledgements

Thanks to those who contributed ideas to this post individually.
Thanks also to [Joao et al 2018](https://www.sciencedirect.com/science/article/pii/S1198743X17307097?via%3Dihub).

