---

title:  "How to call whole genome MLST on the command line"
date:   2023-05-05 9:31:11 -0400
categories: MLST wgMLST cgMLST EToKi ChewBBACA ColorID

---

I have been asked in many different contexts how to call alleles on the command line.
In August 2022, I gave an internal workshop on how to call alleles on the command line
and I thought I would translate that into a blog post.

If you're here to look at classic MLST callers, this isn't that blog post.
If you want to know more about classic MLST callers, please see our paper:  
Page et al 2017, "Comparison of classical multi-locus sequence typing software for next-generation sequencing data." _Microbial Genomics_

# Summary

The three callers that seem to be great on the command line are

* [ChewBBACA](https://github.com/B-UMMI/chewBBACA)
* [EToKi](https://github.com/zheminzhou/EToKi)
* [ColorID](https://github.com/hcdenbakker/colorid)

Each caller has its positives and negatives and so I cannot say that one or the other is clearly the best one over all others.

For each caller, here is what you should expect:

* Will take a few minutes on a bacterial genome
* Output formats
   * tsv profiles file
      * Assembly name, usually in the first column
      * Allele integers, usually in subsequent columns
   * fasta of allele matches (not always)

A lot of my motivation is to find a CLI supplement for BioNumerics's wgMLST caller,
and so I compared results against BioNumerics.
I had received four benchmark datasets from PulseNet and made a scatterplot
of MLST distances between any two genomes in BioNumerics and the other caller.

When looking at the trendline between callers and BioNumerics, 
the linear relationship is outstanding.

| Dataset | Caller | Trendline | R<sup>2 |
|---------|--------|-----------|---------|
| _Salmonella enterica_   | ChewBBACA | `Y = 1.01x - 0.01` | 1 |
| _Salmonella enterica_   | EToKi     | `Y = 1x + 0`       | 1 | 
| _Salmonella enterica_   | ColorID   | `Y = 1.01x - 0.01` | 1 |
|  STEC                   | ChewBBACA | `Y = 1x - 0.01`    | 1 |
|  STEC                   | EToKi     | `Y = 0.99x + 0`    | 1 | 
|  STEC                   | ColorID   | `Y = 0.99x + 0`    | 1 |
| _Campylobacter jejuni_  | ChewBBACA | `Y = 1x + 0`       | 1 |
| _Campylobacter jejuni_  | EToKi     | `Y = 1x + 0`       | 1 | 
| _Campylobacter jejuni_  | ColorID   | `Y = 1.01x + 0`    | 1 |
| _Listeria monocytogenes_| ChewBBACA | `Y = 1x + 0`       | 1 |
| _Listeria monocytogenes_| EToKi     | `Y = 1x + 0`       | 1 | 
| _Listeria monocytogenes_| ColorID   | `Y = 1x + 0`       | 1 |

I also took a random _Campylobacter jejuni_ genome and ran the same genome
assembly through each of the three callers 50 times each.
Even if the caller had the option for multithreading, I kept it at 1 thread each.

![Benchmark times of three callers](/assets/images/wgMLST-benchmark-times.png)

I will give the caveat that ChewBBACA has gone from version 2 to version 3
and so it is possible that would affect this benchmarking.

# Data for running the callers

# ChewBBACA

<https://github.com/B-UMMI/chewBBACA>

## Installation

ChewBBACA can be installed via conda

    conda create -c bioconda -c conda-forge -n chewie "chewbbaca=3.1.2"

## 
