---

title:  "How to call whole genome MLST on the command line"
date:   2023-05-05 9:31:11 -0400
categories: MLST wgMLST cgMLST EToKi ChewBBACA ColorID

---

I have been asked in many different contexts how to call alleles on the command line.
In August 2022, I gave an internal workshop on how to call alleles on the command line
and I thought I would translate that into a blog post.

If you're here to look at classic MLST callers, this isn't that blog post.
If you want to know more about classic MLST callers, please see our paper:  
Page et al 2017, "Comparison of classical multi-locus sequence typing software for next-generation sequencing data." _Microbial Genomics_

_NOTE_: I will give the caveat that ChewBBACA has gone from version 2 to version 3
and so it is possible that things have changed.

# Summary

The three callers that seem to be great on the command line are

* [ChewBBACA](https://github.com/B-UMMI/chewBBACA)
* [EToKi](https://github.com/zheminzhou/EToKi)
* [ColorID](https://github.com/hcdenbakker/colorid)

Each caller has its positives and negatives and so I cannot say that one or the other is clearly the best one over all others.

A lot of my motivation is to find a CLI supplement for BioNumerics's wgMLST caller,
and so I compared results against BioNumerics.
I had received four benchmark datasets from PulseNet and made a scatterplot
of MLST distances between any two genomes in BioNumerics and the other caller.

When looking at the trendline between callers and BioNumerics, 
the linear relationship is outstanding.

| Dataset | Caller | Trendline | R<sup>2 |
|---------|--------|-----------|---------|
| _Salmonella enterica_   | ChewBBACA | `Y = 1.01x - 0.01` | 1 |
| _Salmonella enterica_   | EToKi     | `Y = 1x + 0`       | 1 | 
| _Salmonella enterica_   | ColorID   | `Y = 1.01x - 0.01` | 1 |
|  STEC                   | ChewBBACA | `Y = 1x - 0.01`    | 1 |
|  STEC                   | EToKi     | `Y = 0.99x + 0`    | 1 | 
|  STEC                   | ColorID   | `Y = 0.99x + 0`    | 1 |
| _Campylobacter jejuni_  | ChewBBACA | `Y = 1x + 0`       | 1 |
| _Campylobacter jejuni_  | EToKi     | `Y = 1x + 0`       | 1 | 
| _Campylobacter jejuni_  | ColorID   | `Y = 1.01x + 0`    | 1 |
| _Listeria monocytogenes_| ChewBBACA | `Y = 1x + 0`       | 1 |
| _Listeria monocytogenes_| EToKi     | `Y = 1x + 0`       | 1 | 
| _Listeria monocytogenes_| ColorID   | `Y = 1x + 0`       | 1 |

I also took a random _Campylobacter jejuni_ genome and ran the same genome
assembly through each of the three callers 50 times each.
Even if the caller had the option for multithreading, I kept it at 1 thread each.

![Benchmark times of three callers](/assets/images/wgMLST-benchmark-times.png)

This table is my semi-subjective overall review of the callers.

| Measure            | EToKi                                                                                                                         | ChewBBACA                                                                                              | ColorID                                   |
| ------------------ | ----------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ | ----------------------------------------- |
| Algorithm          | Gene prediction, “uberblast”  to ref alleles to get locus, then hashsum comparison against complete database for allele call. | Gene prediction, BLASTp to ref alleles to get locus, then BLASTp to complete database for allele call. | Colored Debruijn graphs                   |
| Developers         | Single developer; Slightly responsive                                                                                                           | PhD students; responsive                                                                               | Single developer; Very responsive                           |
| Documentation      | Minimal                                                                                                                       | Extensive wiki                                                                                         | Medium                                    |
| Installation       | Source; (Conda); Docker                                                                                                       | Source; Conda; Docker                                                                                  | Cargo (easy)                              |
| Adding new alleles | Not yet; should be easy to add                                                                                                | Automatically done but also alters the database without approval                                       | Cannot be done but could be a new feature |

# Let's call alleles

We'll turn this overview into a workshop starting here.

## Get your data

1. Make a folder to stay organized. The code for this list is below this list.
2. Get a database from <https://chewbbaca.online>.
For this workshop, I am going to use <https://chewbbaca.online/api/NS/api/download/compressed_schemas/4/1/2021-05-30T22:06:50.902917> which is a download link all the way to the right side on [the Campylobacter page](https://chewbbaca.online/species/4).
3. Get an assembly you want to call alleles with. I chose 15AR0919 at random for this blog post since it is a complete genome.

```bash
mkdir -pv ~/workshops/wgMLST
cd ~/workshops/wgMLST
mkdir Campy.chewbbaca
cd Campy.chewbbaca
wget -O Campylobacter_jejuni_INNUENDO_wgMLST_2021-05-30T22_06_50.902917.zip "https://chewbbaca.online/api/NS/api/download/compressed_schemas/4/1/2021-05-30T22 :06:50.902917"
unzip Campylobacter_jejuni_INNUENDO_wgMLST_2021-05-30T22_06_50.902917.zip
cd ..
mkdir asm
# I took this download link from interacting on the NCBI website at https://www.ncbi.nlm.nih.gov/nuccore/NZ_CP035894.1
wget -O asm/15AR0919.fasta "https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&save=file&log$=seqview&db=nuccore&report=fasta&id=1804576668&&ncbi_phid=null"
```

## What to expect when calling alleles

For each caller, here is what you should expect:

* Will take a few minutes on a bacterial genome
* Output formats
   * tsv profiles file
      * Assembly name, usually in the first column
      * Allele integers, usually in subsequent columns
   * fasta of allele matches (not always)

## ChewBBACA

<https://github.com/B-UMMI/chewBBACA>

### ChewBBACA Installation

ChewBBACA can be installed via conda

    conda create -c bioconda -c conda-forge -n chewie "chewbbaca=3.1.2"
    conda activate chewie

## ChewBBACA Calling alleles

Calling alleles is pretty easy and takes about 2 minutes with 8 cpus.
It is efficient in batching and so your times might speed up per genome if you are able to query multiples at once.

    chewie AlleleCall -i asm --schema-directory Campy.chewbbaca -o chewie.out --cpu 8

For chewie version 3, you can add `--no-inferred` to avoid adding new alleles to the database as you are querying.
For chewie version 2, you can add `--fr` to ignore any temporary files in the database and start fresh.

## ChewBBACA Results

    $ ls chewie.out/
    cds_coordinates.tsv  loci_summary_stats.tsv  paralogous_counts.tsv  results_alleles.tsv      results_statistics.tsv
    invalid_cds.txt      logging_info.txt        paralogous_loci.tsv    results_contigsInfo.tsv

